src/main/java/org/candlepin/util/NonNullLinkedHashSet.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2009 - 2024 Red Hat, Inc.
+ *
+ * This software is licensed to you under the GNU General Public License,
+ * version 2 (GPLv2). There is NO WARRANTY for this software, express or
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.util;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+
+/**
+ * An extension of {@link LinkedHashSet} where null values cannot be added.
+ *
+ * @param <E>
+ *  the object type populating the set
+ */
+public class NonNullLinkedHashSet<E> extends LinkedHashSet<E> {
+
+    /**
+     * {@inheritDoc}
+     *
+     * @param object
+     *  non-null element to be added to the set
+     *
+     * @throws IllegalArgumentException
+     *  if the provided element is null
+     */
+    @Override
+    public boolean add(E object) {
+        if (object == null) {
+            throw new IllegalArgumentException("object cannot be null");
+        }
+
+        return super.add(object);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @param collection
+     *  collection to add to the set
+     *
+     * @throws IllegalArgumentException
+     *  if the provided collection contains null
+     */
+    @Override
+    public boolean addAll(Collection<? extends E> collection) {
+        if (collection == null) {
+            return false;
+        }
+
+        if (collection.contains(null)) {
+            throw new IllegalArgumentException("collection cannot contain null");
+        }
+
+        return super.addAll(collection);
+    }
+
+    /**
+     * Compares the elements and determines equality with the provided collection. All elements must be the
+     * same and in the same order for this method to return true.
+     *
+     * @param other
+     *  the collection to check equality for
+     *
+     * @return true if and only if the elements are all equal and in the same order
+     */
+    public boolean equalsCollection(Collection<E> other) {
```

Comment:
With this being an object now, is this method needed? Would using the natural .equals method suffice?

Reply:
Yeah, we should just override the .equals method here. Updated

Reply:
You probably don't even need to override it (or hashcode). The parent class should still handle null values in its equality check, and since we don't allow them, the target collection having nulls should fail it.

Reply:
The problem with using the `AbstractSet` equals method is that it will always return false for non Set instances. This is a problem because `EnvironmentCurator.findEnvironmentsOf` returns a List of environment IDs and we compare this List with a provided NonNullLinkedHashSet in `ConsumerManager.setConsumersEnvironments`. 

We could update `ConsumerManager.setConsumersEnvironments` to return a NonNullLinkedHashSet or Set instead of a List, but then we would have to update `EnvironmentResource.prepareEnvironmentUpdates` and probably `EnvironmentUpdates` as well.

AbstractSet.equals
```
   public boolean equals(Object o) {
      if (o == this) {
         return true;
      } else if (!(o instanceof Set)) {
         return false;
      } else {
         Collection<?> c = (Collection)o;
         if (c.size() != this.size()) {
            return false;
         } else {
            try {
               return this.containsAll(c);
            } catch (NullPointerException | ClassCastException var4) {
               return false;
            }
         }
      }
   }
```

Reply:
Removed the equals and hashcode override and added a private method in ConsumerManager that checks the elements of a `NonNullLinkedHashSet` with the elements in a collection.

Reply:
That's really obnoxious, but I suppose somewhat makes sense. I'd have liked it if it were smart enough to do logical equality, but such is life I suppose.